core/io/logger.cpp

```diff
@@ -266,3 +267,354 @@ CompositeLogger::~CompositeLogger() {
+		// it's not Frame 0, therefore we don't have to mess around with the buffer at all
+		// just lock, register ourselves as a callable, and recalculate the state so we can start capturing messages
+		MutexLock lock(mutex);
+		register_callable(captures_nonthreadsafe, p_callable);
+		recalculate_state();
+	}
+}
+void UserLogManagerLogger::unregister_log_capture_nonthreadsafe(const Callable &p_callable) {
+	// It gets *extremely* hard to guarantee the proper semantics if you're allowed to call this from other threads.
+	ERR_FAIL_COND_MSG(!::Thread::is_main_thread(), "This call is forbidden outside the main thread.");
+
+	MutexLock lock(mutex);
+	unregister_callable(captures_nonthreadsafe, p_callable);
+	recalculate_state();
+}
+void UserLogManagerLogger::register_log_capture_buffered(const Callable &p_callable) {
+	// It gets *extremely* hard to guarantee the proper semantics if you're allowed to call this from other threads.
+	ERR_FAIL_COND_MSG(!::Thread::is_main_thread(), "This call is forbidden outside the main thread.");
+
+	MutexLock lock(mutex);
+	register_callable(captures_buffered, p_callable);
+	recalculate_state();
+}
+void UserLogManagerLogger::unregister_log_capture_buffered(const Callable &p_callable) {
+	// It gets *extremely* hard to guarantee the proper semantics if you're allowed to call this from other threads.
+	ERR_FAIL_COND_MSG(!::Thread::is_main_thread(), "This call is forbidden outside the main thread.");
+
+	MutexLock lock(mutex);
+	unregister_callable(captures_buffered, p_callable);
+	recalculate_state();
+}
+
+void UserLogManagerLogger::flush() {
+	// if you're not sure why this is important, go read the giant comment near the end of recalculate_state
+	// it avoids a nearly-impossible race condition
+	ERR_FAIL_COND_MSG(!::Thread::is_main_thread(), "This call is forbidden outside the main thread.");
+
+	// This flushes our buffer and recycles it for the next frame.
+	// If we don't have a buffer, we have nothing to do.
+	if (state != STATE_BUFFERING) {
+		return;
+	}
+
+	// "Send all our buffered messages to all our buffered log readers"
+	// To avoid inconsistent message delivery of anything that's already been buffered, we swap the entire buffer first
+	// Anything that gets delivered while we're writing these messages gets to wait for next frame.
+	// (We could loop until the buffer is empty)
+	Vector<Dictionary> buffered_logs_mirror;
+	{
+		MutexLock lock(mutex);
+		SWAP(buffered_logs_mirror, buffered_logs);
+	}
+
+	// Dispatch to all the buffered callables, in a threadsafe manner
+	// Any buffered callables that disable themselves stop getting messages ASAP
+	// any buffered callables that get attached might start getting messages midway through
+	// we're ok with that, it's still a chronologically coherent block
+	for (int i = 0; i < buffered_logs_mirror.size(); ++i) {
+		// This is the same index-as-iterator-to-avoid-lock-issues dance we do in register_log_capture_nonthreadsafe()
+		int indexToSendTo = 0;
```

Comment:
Please use snake_case naming for variables as per the [code style guidelines](https://docs.godotengine.org/en/stable/contributing/development/code_style_guidelines.html).

Reply:
Done! I'm honestly surprised the style checks don't catch this one.

Reply:
It might be https://clang.llvm.org/extra/clang-tidy/checks/readability/identifier-naming.html - I may take a look at this once I'm back home.

. . . entirely because I know I'm going to screw this up *all the time* if I don't have guardrails.

Reply:
(gonna close this myself just to clean it up, re-open if there's info you wanted in here)